daemon/algod/api/server/v2/simulate.go

```diff
@@ -0,0 +1,205 @@
+// We override this to use the set hdr to prevent racing with the network
+func (l apiSimulatorLedgerConnector) Latest() basics.Round {
+	return l.hdr.Round
+}
+
+// BlockHdr is part of the LedgerForSimulator interface.
+// We override this to use the set hdr to prevent racing with the network
+func (l apiSimulatorLedgerConnector) BlockHdr(round basics.Round) (bookkeeping.BlockHeader, error) {
+	if round != l.Latest() {
+		err := fmt.Errorf(
+			"BlockHdr() evaluator called this function for the wrong round %d, "+
+				"latest round is %d",
+			round, l.Latest())
+		return bookkeeping.BlockHeader{}, err
+	}
+
+	return l.LedgerForAPI.BlockHdr(round)
+}
+
+// BlockHdrCached is part of the LedgerForSimulator interface.
+func (l apiSimulatorLedgerConnector) BlockHdrCached(round basics.Round) (bookkeeping.BlockHeader, error) {
+	return l.BlockHdr(round)
+}
+
+// GenesisHash is part of LedgerForSimulator interface.
+func (l apiSimulatorLedgerConnector) GenesisHash() crypto.Digest {
+	return l.hdr.GenesisHash
+}
+
+// GenesisProto is part of LedgerForSimulator interface.
+func (l apiSimulatorLedgerConnector) GenesisProto() config.ConsensusParams {
+	return config.Consensus[l.hdr.CurrentProtocol]
+}
+
+// GetCreatorForRound is part of LedgerForSimulator interface.
+func (l apiSimulatorLedgerConnector) GetCreatorForRound(round basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType) (creator basics.Address, ok bool, err error) {
+	if round != l.Latest() {
+		err = fmt.Errorf(
+			"GetCreatorForRound() evaluator called this function for the wrong round %d, "+
+				"latest round is %d",
+			round, l.Latest())
+		return
+	}
+
+	return l.GetCreator(cidx, ctype)
+}
+
+func makeLedgerForSimulatorFromLedgerForAPI(ledgerForAPI LedgerForAPI, hdr bookkeeping.BlockHeader) LedgerForSimulator {
+	return &apiSimulatorLedgerConnector{ledgerForAPI, hdr}
+}
+
+// ==============================
+// > Simulator Errors
+// ==============================
+
+// SimulatorError is the base error type for all simulator errors.
+type SimulatorError struct {
+	error
+}
+
+// InvalidTxGroupError occurs when an invalid transaction group was submitted to the simulator.
+type InvalidTxGroupError struct {
+	SimulatorError
+}
+
+// InvalidSignatureError occurs when a transaction has an invalid signature.
+type InvalidSignatureError struct {
+	SimulatorError
+}
+
+// ScopedSimulatorError is a simulator error that has 2 errors, one for internal use and one for
+// displaying publicly. THe external error is useful for API routes/etc.
+type ScopedSimulatorError struct {
+	SimulatorError        // the original error for internal use
+	External       string // the external error for public use
+}
+
+// ==============================
+// > Simulator Helper Methods
+// ==============================
+
+func isInvalidSignatureError(err error) bool {
+	var signatureError verify.SignatureError
+	if errors.As(err, &signatureError) {
+		return true
+	}
+
+	return false
```

Comment:
```suggestion
	return errors.As(err, &verify.SignatureError{})
```

Reply:
Might as well use `errors.Is` if you don't need to inspect the error instance

Reply:
Then we would need to change how these errors are created.  Each one is a new instance, created by `fmt.Errorf` so there is no instance you can compare them to with Error.Is.

I'm actually getting worried about `SignatureError`.  It's been declared as:
```
type SignatureError interface {
	error
}
```

Doesn't that mean that _any_ error instance can be assigned to one of these (they are, effectively, both the interface that demands `Error()`).  If so, then `Error.As(err, &verify.SignatureError{})` will _always_ return `true`.  Or am I misunderstanding interface semantics.

I think I'm not, because this compiles, implying you are converting a random `error` instance into a SignatureError automatically
```
func signatureErrorf(format string, args ...interface{}) SignatureError {
	return fmt.Errorf(format, args...)
}
```

I think that what you really want to do is create a new `struct` type SIgnatureError that contains an error.

```
type SigError struct {
	err  error
}
func (se SigError) Error() string {
       return se.err.Err() // or, I suppose say something extra
}
```
create them with `SigError{fmt.Errorf()}` or just `SigError{errors.New()}`
```
Then test with `errors.As(err, &SigError{})`

I'm sure I've got something wrong in there.

Reply:
I think you could use the embedded interface trick I mentioned in standup, but just using:
```
type SigError struct {
  error
}
```
without bothering to define an `Error()` on SigError (it'll get promoted).  The key is you need a struct here, not an interface, because embedding an interface in an interface just gives the same interface.

Reply:
Ah I missed that the original `SignatureError` was an interface. +1 to making it a struct.